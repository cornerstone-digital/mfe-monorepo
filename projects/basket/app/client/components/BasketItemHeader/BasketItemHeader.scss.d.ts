export type Styles = {
  'action-button': string
  'action-text': string
  blue: string
  brand: string
  'fade-in': string
  'fade-out': string
  'is-disabled': string
  'is-hidden': string
  'is-upgraded': string
  'item-header': string
  'item-title': string
  'sub-title': string
  'total-cost': string
}

export type ClassNames = keyof Styles

declare const styles: Styles

export default styles
