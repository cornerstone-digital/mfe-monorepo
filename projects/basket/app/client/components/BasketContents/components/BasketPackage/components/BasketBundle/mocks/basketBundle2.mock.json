{
  "basketId": "e5be8241-7b99-4c48-b2b4-d98a2eaec88a",
  "affiliateId": null,
  "affiliateFlag": false,
  "source": "WCS",
  "customerRequestedDate": "2018-11-22T13:11:00.000Z",
  "metadata": {
    "returnURL": null,
    "basketStatus": "INCOMPLETE",
    "salesOrderId": null,
    "reason": null
  },
  "customer": null,
  "journey": null,
  "packages": [{
    "packageId": "684b793c-88b6-4b3f-a4ab-863e26e37f75",
    "accountCategory": "Consumer",
    "sequence": 2,
    "packageType": "Acquisition",
    "removable": true,
    "dependentOn": [],
    "voucherCode": null,
    "bundle": {
      "packageLineId": "94ffbac8-1703-4226-94a8-d5447f296ca7",
      "action": "ADD",
      "skuId": "110546",
      "name": "Gigafast 100 with Voice -18 months contract Unlimited Broadband",
      "description": null,
      "displayName": "Vodafone Gigafast Broadband 100",
      "displayDescription": "Vodafone Gigafast Broadband 100",
      "bundleClass": "HANDSET",
      "productClass": null,
      "serviceId": "107551",
      "paymentType": "Postpaid",
      "bundleType": null,
      "commitmentPeriod": {
        "uom": "Months",
        "value": "18"
      },
      "displayOrder": null,
      "showInSummary": false,
      "allowances": null,
      "merchandisingMedia": [],
      "attributes": null,
      "priceDetails": {
        "monthlyPrice": {
          "gross": "28",
          "net": "23.33",
          "vat": "4.67"
        },
        "monthlyDiscountPrice": {
          "gross": "28",
          "net": "23.33",
          "vat": "4.67"
        },
        "merchandisingPromotions": null
      },
      "isVodafoneGlobalRoaming": false,
      "bundledServiceProducts": [{
        "skuId": "107552",
        "name": "Fixed Broadband Service",
        "description": "Fixed Broadband Service",
        "productClass": "Fixed Broadband Service",
        "servicePointId": "110560",
        "priceDetail": {
          "oneOffPrice": null,
          "oneOffDiscountPrice": null,
          "monthlyPrice": null,
          "monthlyDiscountPrice": null,
          "merchandisingPromotions": null
        },
        "miscellaneous": [{
          "name": "SERVICE POINT",
          "value": "Gigafast 100 Broadband"
        }]
      }, {
        "skuId": "107551",
        "name": "Fixed Line Service",
        "description": "Fixed Line Service",
        "productClass": "Fixed Line",
        "servicePointId": "110567",
        "priceDetail": {
          "oneOffPrice": null,
          "oneOffDiscountPrice": null,
          "monthlyPrice": null,
          "monthlyDiscountPrice": null,
          "merchandisingPromotions": null
        },
        "miscellaneous": [{
          "name": "SERVICE POINT",
          "value": "Gigafast Home Broadband and Voice"
        }]
      }],
      "planCategory": "CONSUMER",
      "isChanged": false
    },
    "hardwares": [{
      "packageLineId": "06b180fe-cc4c-4d2b-b4a5-e283ddcbced7",
      "action": "ADD",
      "skuId": "200692",
      "name": "Vodafone VOX 3.0 Uk",
      "description": null,
      "displayName": "Vodafone Broadband Router",
      "displayDescription": null,
      "deviceBriefDescription": [],
      "deviceColor": null,
      "memorySize": null,
      "make": null,
      "model": null,
      "groupType": null,
      "productClass": "Fixed Broadband Equipment",
      "displayOrder": null,
      "showInSummary": false,
      "merchandisingMedia": [{
        "id": "imageURLs.thumbs.front",
        "value": "/assets/images/desktop/IMG_170x170_VOIX_30_Router.png",
        "type": "URL",
        "footNotes": [""]
      }],
      "attributes": [],
      "priceDetails": {
        "oneOffPrice": {
          "gross": "0",
          "net": "0",
          "vat": "0"
        },
        "oneOffDiscountPrice": {
          "gross": "0",
          "net": "0",
          "vat": "0"
        },
        "merchandisingPromotions": null,
        "monthlyPrice": null,
        "monthlyDiscountPrice": null,
        "financingOptions": [],
        "deviceFinancingId": null
      },
      "boxPrice": {
        "gross": "69.72",
        "net": "58.10",
        "vat": "11.62"
      },
      "freeGift": false,
      "isChanged": false
    }],
    "services": [{
      "packageLineId": "6572d967-0aae-4cad-a032-e65860ec988b",
      "action": "ADD",
      "skuId": "110565",
      "name": "Standard Installation Fee",
      "description": null,
      "displayName": "Standard Installation Fee",
      "displayDescription": null,
      "specificationGroup": [],
      "merchandisingMedia": [],
      "attributes": [],
      "priceDetails": {
        "oneOffPrice": {
          "gross": "0",
          "net": "0",
          "vat": "0"
        },
        "oneOffDiscountPrice": {
          "gross": "0",
          "net": "0",
          "vat": "0"
        },
        "monthlyPrice": null,
        "monthlyDiscountPrice": null,
        "merchandisingPromotions": null
      },
      "freeGift": false,
      "productClass": "Installation Fee",
      "dataRollover": null,
      "rootProductId": null,
      "allowances": null,
      "rewardPoints": null,
      "planDuration": null,
      "isChanged": false
    }, {
      "packageLineId": "6572d967-0aae-4cad-a032-e65860ec988b",
      "action": "ADD",
      "skuId": "110565",
      "name": "Standard",
      "description": null,
      "displayName": "Standard",
      "displayDescription": "Standard",
      "specificationGroup": [],
      "merchandisingMedia": [],
      "attributes": [],
      "priceDetails": {
        "oneOffPrice": {
          "gross": "0",
          "net": "0",
          "vat": "0"
        },
        "oneOffDiscountPrice": {
          "gross": "0",
          "net": "0",
          "vat": "0"
        },
        "monthlyPrice": null,
        "monthlyDiscountPrice": null,
        "merchandisingPromotions": null
      },
      "freeGift": false,
      "productClass": "Installation Fee",
      "dataRollover": null,
      "rootProductId": null,
      "allowances": null,
      "rewardPoints": null,
      "planDuration": null,
      "isChanged": false
    }],
    "discounts": [],
    "priceDetails": {
      "oneOffPrice": {
        "gross": "0",
        "net": "0",
        "vat": "0"
      },
      "oneOffDiscountPrice": {
        "gross": "0",
        "net": "0",
        "vat": "0"
      },
      "monthlyPrice": {
        "gross": "28",
        "net": "23.33",
        "vat": "4.67"
      },
      "monthlyDiscountPrice": {
        "gross": "28",
        "net": "23.33",
        "vat": "4.67"
      },
      "financingOptions": []
    },
    "installationAddress": {
      "citySubDivisionName": null,
      "country": "United Kingdom",
      "county": null,
      "flatNumber": null,
      "houseName": null,
      "houseNumber": "89",
      "locality": null,
      "moveTypeCode": "",
      "postCode": "YO30 5TB",
      "streetName": "EASTHOLME DRIVE",
      "town": "York"
    },
    "appointmentWindow": null,
    "servicePoint": null,
    "serviceStartDate": null,
    "phoneNumber": null,
    "linePackageType": null,
    "lineTreatmentType": null,
    "planType": "BroadBand:FTTH",
    "isChanged": false
  }],
  "vetOutcome": null,
  "priceDetails": {
    "oneOffPrice": {
      "gross": "0",
      "net": "0",
      "vat": "0"
    },
    "oneOffDiscountPrice": {
      "gross": "0",
      "net": "0",
      "vat": "0"
    },
    "monthlyPrice": {
      "gross": "28",
      "net": "23.33",
      "vat": "4.67"
    },
    "monthlyDiscountPrice": {
      "gross": "0",
      "net": "0",
      "vat": "0"
    }
  },
  "totalOneoffPriceSaving": "0.0",
  "totalMonthlyPriceSaving": "84",
  "deliveryInfo": null,
  "voucherCode": null,
  "voucherDescription": null,
  "validated": false,
  "hashvalue": "1003091324",
  "changeCode": null,
  "voucherErrorMessage": null,
  "activeBundles": [],
  "isChanged": false
}
