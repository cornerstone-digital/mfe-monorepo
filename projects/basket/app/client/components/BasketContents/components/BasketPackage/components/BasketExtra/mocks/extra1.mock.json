{
  "packageId": "684b793c-88b6-4b3f-a4ab-863e26e37f75",
  "service": {
    "packageLineId": "6572d967-0aae-4cad-a032-e65860ec988b",
    "action": "ADD",
    "skuId": "110461",
    "name": "Vodafone Insurance: Tier 1 Loss Theft and Damage",
    "description": null,
    "displayName": "Loss, theft, damage and breakdown cover",
    "displayDescription": null,
    "specificationGroup": [],
    "merchandisingMedia": [],
    "attributes": [],
    "priceDetails": {
      "oneOffPrice": null,
      "oneOffDiscountPrice": null,
      "monthlyPrice": {
        "gross": "13.50",
        "net": "13.50",
        "vat": "0"
      },
      "monthlyDiscountPrice": {
        "gross": "13.50",
        "net": "13.50",
        "vat": "0"
      },
      "merchandisingPromotions": null
    },
    "freeGift": false,
    "productClass": "Insurance",
    "dataRollover": null,
    "rootProductId": "110461",
    "allowances": null,
    "rewardPoints": null,
    "planDuration": null,
    "isChanged": false,
    "headerStatus": "removed"
  },
  "planType": "HANDSET",
  "paymentType": "Postpaid"
}