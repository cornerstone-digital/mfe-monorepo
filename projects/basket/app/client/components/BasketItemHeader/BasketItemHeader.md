Plain

```
<BasketItemHeader
  title='Basic header' 
/>
```

Package (present)

```
<BasketItemHeader
  title='45GB Red Entertainment plan' 
  onRemove={() => { alert('remove package') }}
  onUndo={() => { alert('restore package') }}
  headerStatus='present'
  changePackageLink='/'
  isPackage={true}
/>
```

Package (present) with no change button

```
<BasketItemHeader
  showChangeButton={false}
  title='45GB Red Entertainment plan' 
  onRemove={() => { alert('remove package') }}
  onUndo={() => { alert('restore package') }}
  headerStatus='present'
  changePackageLink='/'
  isPackage={true}
/>
```

Package (removing)

```
<BasketItemHeader
  title='45GB Red Entertainment plan'
  headerStatus='removing'
  onRemove={() => { alert('remove package') }}
  onUndo={() => { alert('restore package') }}
  changePackageLink='/'
  isPackage={true}
/>
```

Package (removed)

```
<BasketItemHeader
  title='45GB Red Entertainment plan' 
  onRemove={() => { alert('remove package') }}
  onUndo={() => { alert('restore package') }}
  headerStatus='removed'
  changePackageLink='/'
  isPackage={true}
/>
```

Package (retrieving)

```
<BasketItemHeader
  title='45GB Red Entertainment plan'
  headerStatus='retrieving'
  onRemove={() => { alert('remove package') }}
  onUndo={() => { alert('restore package') }}
  changePackageLink='/'
  isPackage={true}
/>
```

Extra (present)

```
<BasketItemHeader
  title='Loss, theft, damage and breakdown cover' 
  onRemove={() => { alert('remove extra') }}
  onUndo={() => { alert('restore extra') }}
  headerStatus='present'
/>
```

Extra (removing)

```
<BasketItemHeader
  title='Loss, theft, damage and breakdown cover'
  onRemove={() => { alert('remove extra') }}
  onUndo={() => { alert('restore extra') }}
  headerStatus='removing'
/>
```


Extra (removed)

```
<BasketItemHeader
  title='Loss, theft, damage and breakdown cover' 
  onRemove={() => { alert('remove extra') }}
  onUndo={() => { alert('restore extra') }}
  headerStatus='removed'
/>
```

Extra (retrieving)

```
<BasketItemHeader
  title='Loss, theft, damage and breakdown cover'
  onRemove={() => { alert('remove extra') }}
  onUndo={() => { alert('restore extra') }}
  headerStatus='retrieving'
/>
```

Total cost

```
<BasketItemHeader
  title='Total cost'
  isTotalCost={true}
/>
```