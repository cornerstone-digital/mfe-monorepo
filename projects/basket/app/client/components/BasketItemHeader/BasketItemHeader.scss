@import 'node_modules/@vfuk/web-core/resources/resources.scss';

.item-header {
  transition: all 0.3s ease-in;
  opacity: 1;
  display: flex;
  justify-content: space-between;
  align-items: center;

  &.brand {
    color: $color-vodafone-red;
  }

  &.blue {
    color: $color-turquoise;
  }

  &.is-hidden {
    opacity: 0;
  }
}

.action-button {
  display: flex;
  align-items: center;

  &.is-disabled {
    cursor: default;
    outline: none;
  }
}

.action-text {
  display: none;

  @include md {
    margin-bottom: -5px;
    margin-right: $spacing-fixed-1;
    display: inline-block;
  }
}

.item-title a {
  font-weight: normal;
  font-size: $base-font-size;
  margin-left: $spacing-fixed-1;

  @include md {
    font-size: $base-font-size-deskop;
  }
}

.total-cost {
  margin-top: 0;
}

.sub-title {
  color: $color-vodafone-red;
  font-weight: bold;

  &.is-upgraded {
    color: $color-turquoise;
  }
}
