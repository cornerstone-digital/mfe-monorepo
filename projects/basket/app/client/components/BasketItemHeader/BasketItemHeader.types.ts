import { DataSpeedStatus } from '@pages/BasketPage/BasketPage.types'

export type HeaderStatusType = 'present' | 'removing' | 'removed' | 'retrieving'

export interface BasketItemHeaderProps {
  /** the title to show in the header */
  title?: string
  /** the color of title to show in the header */
  titleColor?: 'brand' | 'blue'
  /** the sub title to show in the header */
  subTitle?: string
  /** the event to fire when clicking the remove action */
  onRemove?: () => void
  /** the event to fire when clicking the undo action */
  onUndo?: () => void
  /** the current state of the header */
  headerStatus?: HeaderStatusType
  /** specify whether this is a package */
  isPackage?: boolean
  /** specify whether this is an upgrade */
  isUpgrade?: boolean
  /** include the change link */
  changePackageLink?: string
  /** specify if this is a special total cost item header */
  isTotalCost?: boolean
  /** Whether to show or remove the change button, defaults to true */
  showChangeButton?: boolean

  /** relevant package ID */
  packageId?: string
  /** for analytics */
  pageError?: string
  /** review mode */
  reviewMode?: boolean
  actionDisabled?: boolean
  dataSpeed?: {
    key: DataSpeedStatus
    message?: string
  }
}

export type HeaderStatus = {
  headerTitle?: string
  headerSubTitle?: string
  headerAction: string
  headerIcon: JSX.Element
}

export interface HeaderStatuses {
  present: HeaderStatus
  removing: HeaderStatus
  removed: HeaderStatus
  retrieving: HeaderStatus
}

export interface HeaderClassNames {
  containerClassNames: string
  actionButtonClassNames: string
  subTitleClassNames: string
}
