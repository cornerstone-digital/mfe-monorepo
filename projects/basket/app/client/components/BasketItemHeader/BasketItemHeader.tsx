import { useEffect, useRef, useContext, useState } from 'react'
import classnames from 'classnames'

import Heading from '@web-core/components/atoms/Heading'
import Paragraph from '@web-core/components/atoms/Paragraph'

import ChangeLink from './internal/ChangeLink'
import BasketContentsContext from '@components/BasketContents/context/BasketContentsContext'

import getHeaderStatusDetails from './helpers/getHeaderStatusDetails'
import DataSpeedButton from './internal/DataSpeedButton'
import getItemType from './helpers/getItemType'
import getABTestFeatureValue from '@helpers/getABTestFeatureValue'

import styles from './BasketItemHeader.scss'
import { BasketItemHeaderProps, HeaderClassNames, HeaderStatusType } from './BasketItemHeader.types'
import getSelectorFromTitle from './helpers/getSelectorFromHeaderTitle'

const BasketItemHeader = (props: BasketItemHeaderProps) => {
  const {
    headerStatus = 'present',
    isPackage = false,
    showChangeButton = true,
    onRemove,
    onUndo,
    isTotalCost,
    titleColor,
    isUpgrade,
    changePackageLink,
    title,
    subTitle,
    pageError,
    dataSpeed,
    packageId,
    actionDisabled,
  } = props

  const [isTransitioning, setIsTransitioning] = useState(false)
  const [savedHeaderStatus, setHeaderStatus] = useState<HeaderStatusType>(headerStatus)
  const { isReviewMode } = useContext(BasketContentsContext)

  let headerAnimateTimeout = useRef<NodeJS.Timeout>()

  const cleanTimeouts = () => {
    if (headerAnimateTimeout.current) {
      clearTimeout(headerAnimateTimeout.current)
    }
  }

  useEffect(() => {
    if (headerStatus !== savedHeaderStatus) {
      cleanTimeouts()
      setIsTransitioning(true)
      headerAnimateTimeout.current = setTimeout(() => {
        setIsTransitioning(false)
        setHeaderStatus(headerStatus)
      }, 300)
    }

    return () => {
      cleanTimeouts()
    }
  }, [headerStatus])

  const handleClick = () => {
    if (onRemove && onUndo && !isTransitioning && ['present', 'removed'].includes(savedHeaderStatus)) {
      if (savedHeaderStatus === 'present') {
        return onRemove && onRemove()
      }
      return onUndo && onUndo()
    }
  }
  const headerActionIsDisabled = ['removing', 'retrieving'].includes(savedHeaderStatus)
  const classNames: HeaderClassNames = {
    containerClassNames: classnames('item-header', {
      'is-hidden': isTransitioning,
      'total-cost': isTotalCost,
      [titleColor as string]: !!titleColor,
    }),
    actionButtonClassNames: classnames('action-button', { 'is-disabled': headerActionIsDisabled }),
    subTitleClassNames: classnames('sub-title', { 'is-upgraded': isUpgrade }),
  }

  const showAction = !!onRemove && !!onUndo && !actionDisabled
  const { containerClassNames, actionButtonClassNames, subTitleClassNames } = classNames

  const showChangeLink =
    showChangeButton &&
    (getABTestFeatureValue('changeLinkAB') || isPackage || showAction) &&
    changePackageLink &&
    savedHeaderStatus === 'present'
  const headerStatusDetails = getHeaderStatusDetails(title, subTitle, savedHeaderStatus, isPackage)
  const headerActionSelector = `basket-header-action-${getSelectorFromTitle(headerStatusDetails?.headerTitle || '')}`

  return (
    <div styleName={containerClassNames}>
      <div className={styles['item-title']}>
        <Heading level={5} size={isTotalCost ? 4 : 6} margin={0} fontWeight="bold">
          {headerStatusDetails?.headerTitle}
          {showChangeLink && !isReviewMode && (
            <ChangeLink
              packageId={packageId}
              pageError={pageError}
              changePackageLink={changePackageLink}
              itemType={getItemType(changePackageLink)}
            />
          )}
        </Heading>
        {dataSpeed?.key && <DataSpeedButton packageId={packageId} pageError={pageError} dataSpeed={dataSpeed} />}
        {headerStatusDetails?.headerSubTitle && (
          <Paragraph fontSize="xs" styleName={subTitleClassNames}>
            {headerStatusDetails.headerSubTitle}
          </Paragraph>
        )}
      </div>
      {showAction && !isReviewMode && (
        <button onClick={handleClick} styleName={actionButtonClassNames} data-selector={headerActionSelector}>
          <span className={styles['action-text']}>{headerStatusDetails?.headerAction}</span>
          {headerStatusDetails?.headerIcon}
        </button>
      )}
    </div>
  )
}
export default BasketItemHeader
