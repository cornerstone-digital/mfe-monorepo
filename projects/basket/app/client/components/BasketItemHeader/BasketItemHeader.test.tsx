import { render, screen } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

import { DataSpeedStatus } from '@pages/BasketPage/BasketPage.types'
import BasketItemHeader from './BasketItemHeader'

import AnalyticsUtil from '@utilities/Analytics'
import { BasketItemHeaderProps } from './BasketItemHeader.types'
import { mockStore } from '@tools/mocks/storeMock'

jest.mock('@utilities/Analytics')

jest.mock('@store', () => ({
  useStore: jest.fn().mockImplementation(() =>
    mockStore({
      basketId: 'some-basket',
    }),
  ),
}))

AnalyticsUtil.trackLink = jest.fn()

const props: BasketItemHeaderProps = {
  title: 'header title',
  onRemove: jest.fn(),
  onUndo: jest.fn(),
  headerStatus: 'present',
  isPackage: true,
  changePackageLink: '',
  packageId: '123',
}

describe('<BasketItemHeader />', () => {
  test('should render title', () => {
    render(<BasketItemHeader {...props} />)
    expect(screen.getByText('header title')).toBeInTheDocument()
  })

  test('should render subtitle', () => {
    render(<BasketItemHeader {...props} subTitle="header subtitle" />)
    expect(screen.getByText('header subtitle')).toBeInTheDocument()
  })

  test('should render change button', () => {
    const { getByText } = render(<BasketItemHeader {...props} changePackageLink="test" showChangeButton />)
    const button = getByText('Change')
    expect(button).toBeInTheDocument()
    expect(button).toHaveClass('change-button')
  })

  test('should render change button for item in combi package', () => {
    const { getByText } = render(<BasketItemHeader {...props} changePackageLink="test" isPackage={false} showChangeButton />)
    const button = getByText('Change')
    expect(button).toBeInTheDocument()
    expect(button).toHaveClass('change-button')
  })

  test('should render action button', () => {
    const { getByText } = render(<BasketItemHeader {...props} />)
    expect(getByText('Remove')).toBeInTheDocument()
    expect(getByText('Remove').parentElement?.tagName.toLowerCase()).toBe('button')
  })

  test('should trigger removal on action button when header status is "present"', () => {
    const removeMock = jest.fn()
    const { getByText } = render(<BasketItemHeader {...props} onRemove={removeMock} />)
    getByText('Remove').click()
    expect(removeMock).toHaveBeenCalled()
  })

  test('should render package header with remove button', () => {
    const { asFragment } = render(<BasketItemHeader {...props} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('should trigger undo on action button when header status is "removed"', () => {
    const undoMock = jest.fn()
    const { getByText } = render(<BasketItemHeader {...props} onUndo={undoMock} headerStatus="removed" />)
    getByText('Undo').click()
    expect(undoMock).toHaveBeenCalled()
  })

  test('should not show remove button during transition', () => {
    const { getByText, queryByText } = render(<BasketItemHeader {...props} headerStatus="removing" />)
    expect(getByText('Removing header title')).toBeInTheDocument()
    expect(queryByText('remove')).toBeNull()
  })

  test('should render removing with loading icon', () => {
    const { asFragment, getByText } = render(<BasketItemHeader {...props} headerStatus="removing" />)
    expect(getByText('Removing header title')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render with undo button', () => {
    const { asFragment, getByText } = render(<BasketItemHeader {...props} headerStatus="removed" />)
    expect(getByText('You removed header title')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render retrieving with loading icon', () => {
    const { asFragment, getByText } = render(<BasketItemHeader {...props} headerStatus="retrieving" />)
    expect(getByText('Retrieving header title')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render extra header', () => {
    const { asFragment, getByText } = render(<BasketItemHeader {...props} isPackage={false} />)
    expect(getByText('header title')).toBeInTheDocument()
    expect(getByText('Remove')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  test('should update correctly', () => {
    const { asFragment, getByText } = render(<BasketItemHeader {...props} headerStatus="removing" isPackage={false} />)
    expect(getByText('Removing header title')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render package header', () => {
    const oldWindow = global.window
    Object.defineProperty(window, 'location', {
      value: {
        href: '',
      },
    })
    const { getByText } = render(<BasketItemHeader {...props} changePackageLink="/basket" showChangeButton />)
    getByText('Change').click()
    expect(AnalyticsUtil.trackLink).toBeCalledWith('basketPage.changePackageCta', {
      itemType: undefined,
      pageError: undefined,
      newBasket: {
        basketId: 'some-basket',
      },
      packageId: '123',
    })
    expect(global.window.location.href).toBe('/basket')

    global.window = oldWindow
  })

  test('should be hidden if transitioning', async () => {
    jest.useFakeTimers()
    const { container, rerender } = render(<BasketItemHeader {...props} />)
    rerender(<BasketItemHeader {...props} headerStatus="removing" />)
    expect(container.querySelector('.vfuk-BasketItemHeader__is-hidden')).toBeInTheDocument()
    act(() => {
      jest.runAllTimers()
    })
    expect(container.querySelector('.vfuk-BasketItemHeader__is-hidden')).toBeNull()
  })

  describe('data speed tests', () => {
    test('should render data speed link', () => {
      props.dataSpeed = {
        key: 'MAX_10_MBPS',
        message: 'Speed: Maximum download of 10 Mbps',
      }
      const { asFragment, getByText } = render(
        <BasketItemHeader
          {...props}
          dataSpeed={{
            key: 'MAX_10_MBPS',
            message: 'Speed: Maximum download of 10 Mbps',
          }}
        />,
      )
      expect(getByText('Speed: Maximum download of 10 Mbps')).toBeInTheDocument()
      expect(getByText('Remove')).toBeInTheDocument()
      expect(asFragment()).toMatchSnapshot()
    })

    describe('when data speed link is clicked', () => {
      test('should render SpeedInfoModal', () => {
        props.dataSpeed = {
          key: 'MAX_10_MBPS',
          message: 'Speed: Maximum download of 10 Mbps',
        }
        const wrapper = shallow(<BasketItemHeader {...props} />)
          .find('DataSpeedButton')
          .dive()
        const dataSpeedButton = wrapper.find('button').at(0)
        dataSpeedButton.find('button').simulate('click')
        expect(wrapper.find('SpeedInfoModal').length).toEqual(1)
      })

      test.each([
        ['MAX_2_MBPS', 'basketPage.dataSpeed2Mbps'],
        ['MAX_10_MBPS', 'basketPage.dataSpeed10Mbps'],
        ['MAX_AVAILABLE', 'basketPage.dataSpeedFastest'],
      ])('should trigger the correct analytics event for different speeds', (speed, expectedAnalyticsEvent) => {
        props.dataSpeed = {
          key: speed as DataSpeedStatus,
          message: 'Speed: Maximum download of Some Mbps',
        }
        const wrapper = shallow(<BasketItemHeader {...props} />)
          .find('DataSpeedButton')
          .dive()
        const dataSpeedButton = wrapper.find('button').at(0)
        dataSpeedButton.find('button').simulate('click')
        expect(AnalyticsUtil.trackLink).toHaveBeenCalledWith(expectedAnalyticsEvent, {
          newBasket: { basketId: 'some-basket' },
          packageId: '123',
        })
      })
    })

    describe('when data speed modal close button is clicked', () => {
      test('should not render SpeedInfoModal', () => {
        props.dataSpeed = {
          key: 'MAX_10_MBPS',
          message: 'Speed: Maximum download of 10 Mbps',
        }
        const wrapper = shallow(<BasketItemHeader {...props} />)
          .find('DataSpeedButton')
          .dive()
        const dataSpeedButton = wrapper.find('button').at(0)
        dataSpeedButton.find('button').simulate('click')
        expect(wrapper.find('SpeedInfoModal').length).toEqual(1)
        const speedinfoModal = wrapper.find('SpeedInfoModal')
        speedinfoModal.prop('onClose')()
        expect(wrapper.find('SpeedInfoModal').length).toEqual(0)
      })
    })
  })
})
