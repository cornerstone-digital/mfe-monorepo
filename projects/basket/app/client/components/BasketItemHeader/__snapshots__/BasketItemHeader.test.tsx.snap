// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BasketItemHeader /> data speed tests should render data speed link 1`] = `
<DocumentFragment>
  <div
    stylename="item-header"
  >
    <div
      class="item-title"
    >
      <h5
        class="heading heading-6 inherit w-bold margin-0"
      >
        header title
      </h5>
      <button
        class="speed-button"
        data-selector="data-speed"
      >
        Speed: Maximum download of 10 Mbps
      </button>
    </div>
    <button
      data-selector="basket-header-action-header-title"
      stylename="action-button"
    >
      <span
        class="action-text"
      >
        Remove
      </span>
      <span
        aria-hidden="true"
        class="wrapper inherit size-2 2  "
      >
        test-file-stub
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<BasketItemHeader /> should render extra header 1`] = `
<DocumentFragment>
  <div
    stylename="item-header"
  >
    <div
      class="item-title"
    >
      <h5
        class="heading heading-6 inherit w-bold margin-0"
      >
        header title
      </h5>
    </div>
    <button
      data-selector="basket-header-action-header-title"
      stylename="action-button"
    >
      <span
        class="action-text"
      >
        Remove
      </span>
      <span
        aria-hidden="true"
        class="wrapper inherit size-1 1  "
      >
        test-file-stub
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<BasketItemHeader /> should render package header with remove button 1`] = `
<DocumentFragment>
  <div
    stylename="item-header"
  >
    <div
      class="item-title"
    >
      <h5
        class="heading heading-6 inherit w-bold margin-0"
      >
        header title
      </h5>
    </div>
    <button
      data-selector="basket-header-action-header-title"
      stylename="action-button"
    >
      <span
        class="action-text"
      >
        Remove
      </span>
      <span
        aria-hidden="true"
        class="wrapper inherit size-2 2  "
      >
        test-file-stub
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<BasketItemHeader /> should render removing with loading icon 1`] = `
<DocumentFragment>
  <div
    stylename="item-header"
  >
    <div
      class="item-title"
    >
      <h5
        class="heading heading-6 inherit w-bold margin-0"
      >
        Removing header title
      </h5>
    </div>
    <button
      data-selector="basket-header-action-removing-header-title"
      stylename="action-button is-disabled"
    >
      <span
        class="action-text"
      />
      <div
        class="wrapper wrapper-left "
      >
        <div
          class="spinner-wrapper spinner-wrapper-left"
        >
          <div
            class="spinner spinner-dark spinner-sm spinner "
          >
            <svg
              viewBox="25 25 50 50"
            >
              <circle
                cx="50"
                cy="50"
                r="20"
              />
            </svg>
            <div
              class="spinner spinner-dark spinner-sm spinner vfuk-LoadingSpinner__fallback_loader"
            >
              <div />
              <div />
              <div />
              <div />
            </div>
          </div>
        </div>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<BasketItemHeader /> should render retrieving with loading icon 1`] = `
<DocumentFragment>
  <div
    stylename="item-header"
  >
    <div
      class="item-title"
    >
      <h5
        class="heading heading-6 inherit w-bold margin-0"
      >
        Retrieving header title
      </h5>
    </div>
    <button
      data-selector="basket-header-action-retrieving-header-title"
      stylename="action-button is-disabled"
    >
      <span
        class="action-text"
      />
      <div
        class="wrapper wrapper-left "
      >
        <div
          class="spinner-wrapper spinner-wrapper-left"
        >
          <div
            class="spinner spinner-dark spinner-sm spinner "
          >
            <svg
              viewBox="25 25 50 50"
            >
              <circle
                cx="50"
                cy="50"
                r="20"
              />
            </svg>
            <div
              class="spinner spinner-dark spinner-sm spinner vfuk-LoadingSpinner__fallback_loader"
            >
              <div />
              <div />
              <div />
              <div />
            </div>
          </div>
        </div>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<BasketItemHeader /> should render with undo button 1`] = `
<DocumentFragment>
  <div
    stylename="item-header"
  >
    <div
      class="item-title"
    >
      <h5
        class="heading heading-6 inherit w-bold margin-0"
      >
        You removed header title
      </h5>
    </div>
    <button
      data-selector="basket-header-action-you-removed-header-title"
      stylename="action-button"
    >
      <span
        class="action-text"
      >
        Undo
      </span>
      <span
        aria-hidden="true"
        class="wrapper inherit size-2 2  "
      >
        test-file-stub
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<BasketItemHeader /> should update correctly 1`] = `
<DocumentFragment>
  <div
    stylename="item-header"
  >
    <div
      class="item-title"
    >
      <h5
        class="heading heading-6 inherit w-bold margin-0"
      >
        Removing header title
      </h5>
    </div>
    <button
      data-selector="basket-header-action-removing-header-title"
      stylename="action-button is-disabled"
    >
      <span
        class="action-text"
      />
      <div
        class="wrapper wrapper-left "
      >
        <div
          class="spinner-wrapper spinner-wrapper-left"
        >
          <div
            class="spinner spinner-dark spinner-xs spinner "
          >
            <svg
              viewBox="25 25 50 50"
            >
              <circle
                cx="50"
                cy="50"
                r="20"
              />
            </svg>
            <div
              class="spinner spinner-dark spinner-xs spinner vfuk-LoadingSpinner__fallback_loader"
            >
              <div />
              <div />
              <div />
              <div />
            </div>
          </div>
        </div>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
